tests/fork-change-slot.c

```diff
@@ -0,0 +1,347 @@
+            argv[0]);
+}
+
+int main(int argc, char *argv[])
+{
+    const EVP_MD *digest_algo = NULL;
+    EVP_PKEY *privkey = NULL, *pubkey = NULL;
+    EVP_MD_CTX *md_ctx = NULL;
+    ENGINE *engine = NULL;
+    X509 *cert = NULL;
+    unsigned char *random = NULL, *signature = NULL;
+
+    int rc = 0, fd, num_processes = 0;
+    unsigned int siglen;
+    pid_t pid;
+
+    int rv = 1;
+
+    /* Check arguments */
+    if (argc < 4) {
+        fprintf(stderr, "Missing required arguments\n");
+        usage(argv);
+        goto failed;
+    }
+
+    if (argc > 5) {
+        fprintf(stderr, "Too many arguments\n");
+        usage(argv);
+        goto failed;
+    }
+
+    /* Check PKCS#11 URL */
+    if (strncmp(argv[1], "pkcs11:", 7)) {
+        fprintf(stderr, "fatal: invalid PKCS#11 URL\n");
+        usage(argv);
+        goto failed;
+    }
+
+    /* Get the number of processes to be spawned */
+    num_processes = atoi(argv[3]);
+    if ((num_processes <= 0) || (num_processes > 100)) {
+        fprintf(stderr, "fatal: invalid number of processes (use from 1 to 100)\n");
+        goto failed;
+    }
+
+    ENGINE_load_builtin_engines();
+
+    pid = getpid();
+
+    printf("pid %d is the parent\n", pid);
+
+    /* Read certificate file */
+    cert = read_cert(argv[2], pid);
+    if (cert == NULL) {
+        error_queue("read_cert", pid);
+        goto failed;
+    }
+
+    /* Get the public key for verification */
+    pubkey = X509_get_pubkey(cert);
+    if (pubkey == NULL) {
+        fprintf(stderr, "pid %d: Could not extract the public key\n", pid);
+        error_queue("X509_get_pubkey", pid);
+        goto failed;
+    }
+
+    /* Get structural reference */
+    engine = ENGINE_by_id("pkcs11");
+    if (engine == NULL) {
+        fprintf(stderr, "fatal: engine \"pkcs11\" not available\n");
+        error_queue("ENGINE_by_id", pid);
+        goto failed;
+    }
+
+    /* Initialize to get the engine functional reference */
+    if (ENGINE_init(engine)) {
+        //TODO Improve arguments parsing using getopt or argp
+        if (argc == 5) {
+            ENGINE_ctrl_cmd(engine, "MODULE_PATH", 0, argv[4], NULL, 1);
+        }
+
+        privkey = ENGINE_load_private_key(engine, argv[1], 0, 0);
+        if (privkey == NULL) {
+            error_queue("ENGINE_load_private_key", pid);
+            goto failed;
+        }
+
+        if (!ENGINE_set_default(engine, ENGINE_METHOD_ALL)) {
+            error_queue("ENGINE_set_default", pid);
+            goto failed;
+        }
+
+        ENGINE_free(engine);
+        engine = NULL;
+    }
+    else {
+        error_queue("ENGINE_init", pid);
+        goto failed;
+    }
+
+    /* Spawn processes and check child return */
+    if (spawn_processes(num_processes)) {
+        goto failed;
+    }
+    pid = getpid();
+
+    /* Get random bytes */
+    random = OPENSSL_malloc(RANDOM_SIZE);
+    if (random == NULL)
+        goto failed;
+
+    fd = open(RANDOM_SOURCE, O_RDONLY);
```

Comment:
Why use /dev/urandom instead of RAND_bytes?

Reply:
Same reason as above.